core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/DefaultLoadBalancingPolicy.java

```diff
@@ -318,4 +330,42 @@ protected int getInFlight(@NonNull Node node, @NonNull Session session) {
     // processing them).
     return (pool == null) ? 0 : pool.getInFlight();
   }
+
+  // Exposed as protected for unit tests
+  protected class NodeResponseRateSample {
+    // The array stores at most two timestamps, since we don't need more;
+    // the first one is always the least recent one, and hence the one to inspect.
+    protected AtomicLongArray times;
+
+    private NodeResponseRateSample() {
+      times = new AtomicLongArray(1);
+      times.set(0, nanoTime());
+    }
+
+    // Only for unit tests
+    protected NodeResponseRateSample(AtomicLongArray times) {
+      this.times = times;
+    }
+
+    private void update() {
+      long now = nanoTime();
+      if (times.length() == 1) {
+        long previous = times.get(0);
+        times = new AtomicLongArray(2);
+        times.set(0, previous);
+        times.set(1, now);
+      } else {
+        times.set(0, times.get(1));
+        times.set(1, now);
+      }
+    }
```

Comment:
This seems to be subject to race conditions to me, although I haven't been able to construct a specific scenario that triggers it.  The old implementation computed a result array locally and then returned that; there was no changing of state at the ultimate destination (the old responseTimes map) beyond updating an entry or not doing so.  This implementation is doing more than that; we're actually changing the types of elements (this.times changes in size) based on data we receive.

It might be okay... but then again I'm also not sure I see a harm in just making this.times be an array of size two from the beginning.

Reply:
As I stated previously, this will race indeed, _unless_ calls to `update()` are externally synchronized. I think the loading cache approach does guarantee that.

The idea of using an array of size 2 from the beginning could save us one array allocation – why not.

Reply:
Comments above refer to an earlier implementation of this functionality.  New version wraps everything in compute() and drives updates off of immutable NodeResponseRateSample objects so these concerns have been addressed.